docs/topics/using-builders-with-builder-inference.md

```diff
@@ -0,0 +1,253 @@
+call available extensions, and access available properties.
+
+### Contributing to builder inference results
+
+There are multiple options to contribute to builder inference results:
+* Calling methods on a lambda's receiver that use a type parameter's type:
+  ```kotlin
+  val result = buildList {
+      // Type argument is inferred into String based on the passed "value" argument
+      add("value")
+  } // result has the List<String> type inferred
+  ```
+* Specifying the expected type for calls which return type parameter's type:
+  ```kotlin
+  val result = buildList {
+      // Type argument is inferred into Float based on the expected type
+      val x: Float = get(0)
+  } // result has the List<Float> type
+  ```
+  ```kotlin
+  class Foo<T> {
+      val items = mutableListOf<T>()
+  }
+
+  fun <K> myBuilder(builder: Foo<K>.() -> Unit): Foo<K> = Foo<K>().apply(builder)
+
+  fun main() {
+      val result = myBuilder {
+          val x: List<CharSequence> = items
+          // ...
+      } // result has the Foo<CharSequence> type
+  }
+  ```
+* Passing postponed type variables' types into methods which expect concrete types:
+  ```kotlin
+  fun takeMyLong(x: Long) { ... }
+
+  fun String.isMoreThat3() = length > 3
+
+  fun takeListOfStrings(x: List<String>) { ... }
+
+  fun main() {
+      val result1 = buildList {
+          val x = get(0)
+          takeMyLong(x)
+      } // result1 has the List<Long> type
+
+      val result2 = buildList {
+          val x = get(0)
+          val isLong = x.isMoreThat3()
+      // ...
+      } // result2 has the List<String> type
+  
+      val result3 = buildList {
+          takeListOfStrings(this)
+      } // result3 has the List<String> type
+  }
+  ```
+* Taking a callable reference to lambda receiver's member:
+  ```kotlin
+  fun main() {
+      val result = buildList {
+          val x: KFunction1<Int, Float> = ::get
+      } // result has the List<Float> type
+  }
+  ```
+  ```kotlin
+  fun takeFunction(x: KFunction1<Int, Float>) { ... }
+
+  fun main() {
+      val result = buildList {
+          takeFunction(::get)
+      } // result has the List<Float> type
+  }
+  ```
+* Using nested interdependent builder inference calls.
+
+Note that the Kotlin compiler uses builder inference only if regular type inference cannot infer a type argument.
+It means you can contribute a type information outside a builder lambda, and then builder inference analysis is not
+required. Consider the example:
+
+```kotlin
+fun someMap() = mutableMapOf<CharSequence, String>()
+
+fun <E> MutableMap<E, String>.f(x: MutableMap<E, String>) { ... }
+
+fun main() {
+    val x: Map<in String, String> = buildMap {
+        put("", "")
+        f(someMap()) // Type mismatch (required String, found CharSequence)
+        // Type mismatch appears because builder inference didn't work here,
+        // the expected Map<in String, String> type is specified outside a builder lambda
```

Comment:
For clarity, we can add "...and builder lambda is being analysed with already fixed receiver's type"

Reply:
[See the commit.](https://github.com/JetBrains/kotlin-web-site/pull/2621/commits/63cc850b733dd2fccf8b250c9064ce0bf3628b49)